library/core/src/ptr/non_null.rs

```diff
@@ -855,6 +862,7 @@ impl<T: ?Sized> NonNull<T> {
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
     #[unstable(feature = "ptr_sub_ptr", issue = "95892")]
     #[rustc_const_unstable(feature = "const_ptr_sub_ptr", issue = "95892")]
+    #[ensures(|result: &usize| *result == self.as_ptr().offset_from(subtracted.as_ptr()) as usize)]
```

Comment:
Can you add safety preconditions? https://doc.rust-lang.org/beta/std/ptr/struct.NonNull.html#method.sub_ptr